test/realtime/repo_test.exs

```diff
@@ -146,40 +146,82 @@ defmodule Realtime.RepoTest do
       timestamp = NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
 
-      result = %Postgrex.Result{
-        columns: ["id", "name", "updated_at", "inserted_at"],
-        rows: [[1, "foo", timestamp, timestamp], [2, "bar", timestamp, timestamp]]
-      }
-
-      metadata = %Ecto.Schema.Metadata{
-        prefix: Channel.__schema__(:prefix),
-        schema: Channel,
-        source: Channel.__schema__(:source),
-        state: :loaded
-      }
-
-      expected = [
-        %Channel{
-          __meta__: metadata,
-          id: 1,
-          name: "foo",
-          updated_at: timestamp,
-          inserted_at: timestamp
-        },
-        %Channel{
-          __meta__: metadata,
-          id: 2,
-          name: "bar",
-          updated_at: timestamp,
-          inserted_at: timestamp
-        }
-      ]
-
-      assert Repo.pg_result_to_struct(result, Channel) == expected
+      result =
+        {:ok,
+         %Postgrex.Result{
+           columns: ["id", "name", "updated_at", "inserted_at"],
+           rows: [[1, "foo", timestamp, timestamp]]
+         }}
+
+      metadata = metadata()
+
+      assert {:ok,
+              %Channel{
+                __meta__: ^metadata,
+                id: 1,
+                name: "foo",
+                updated_at: ^timestamp,
+                inserted_at: ^timestamp
+              }} = Repo.result_to_single_struct(result, Channel)
+    end
+
+    test "no results returns nil" do
+      result =
+        {:ok, %Postgrex.Result{columns: ["id", "name", "updated_at", "inserted_at"], rows: []}}
+
+      assert {:ok, nil} = Repo.result_to_single_struct(result, Channel)
+    end
+
+    test "Postgrex.Result with more than one row will return error" do
+      timestamp = NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
+
+      result =
+        {:ok,
+         %Postgrex.Result{
+           columns: ["id", "name", "updated_at", "inserted_at"],
+           rows: [[1, "foo", timestamp, timestamp], [2, "bar", timestamp, timestamp]]
+         }}
+
+      assert_raise RuntimeError, "expected at most one result but got 2 in result", fn ->
+        Repo.result_to_single_struct(result, Channel)
+      end
+    end
+  end
+
+  describe "result_to_structs/2" do
+    test "converts Postgrex.Result to struct" do
+      timestamp = NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)
+
+      result =
+        {:ok,
+         %Postgrex.Result{
+           columns: ["id", "name", "updated_at", "inserted_at"],
+           rows: [[1, "foo", timestamp, timestamp], [2, "bar", timestamp, timestamp]]
+         }}
+
+      metadata = metadata()
+
+      assert {:ok,
+              [
+                %Channel{
+                  __meta__: ^metadata,
+                  id: 1,
+                  name: "foo",
+                  updated_at: ^timestamp,
+                  inserted_at: ^timestamp
+                },
+                %Channel{
+                  __meta__: ^metadata,
```

Comment:
Instead of pattern matching on the metadata internals, consider assigning `%Channel{}` to a variable and then using `Ecto.get_meta` to access a value of a specific metadata field.

Reply:
will do 👍